base/math.jl

```diff
@@ -536,38 +536,56 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+	#= Keep specialized two argument version, because it is faster
+	than generic varargs version below =#
+	ay, ax = minmax(abs(x), abs(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
     rr = ax * sqrt(1 + r * r)
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
-    if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
-    else
-        return rr
-    end
+	if isnan(r)
+		if isinf(x) || isinf(y)
+			return oftype(rr, Inf)
+		else
+			return oftype(rr, NaN)
+		end
+   	else
+   		return rr
+   	end
 end
-
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = _hypot(promote(x...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
+	N == 0 && error("invalid hypot() call without arguments")
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	if isnan(s)
+		return any(isinf, x) ? oftype(s, Inf) : s # IEEE 754
+	elseif isinf(s) || s ≤ _floatmin(s) * (N-1) # FIXME: revisit when #30387 is resolved
+		# if overflow/underflow, try normalization
+		ma = maximum(abs, x)
+		if iszero(ma) || isinf(ma)
+			return oftype(s, ma)
+		else
+			return oftype(s, ma * sqrt(sum(abs2, x ./ ma)))
```

Comment:
Again, `x ./ ma` allocates a temporary.   Better to do `sum(y -> abs2(y / ma), x)` so that there is a single loop.

To work around #15276, not sure if we need to wrap this in a `let` block.  i.e. `let ma = maximum(abs, x) ....`, to get the best performance?   (Nowadays #15276 doesn't seem to affect so many cases, but it would be good to check.)

Reply:
I think we can just use the dotted calls here. Tuples do not get allocated. Also `@btime` suggests that this way is faster.

Reply:
Or, pursuant to your comment above, `sum(abs2.(x ./ ma))`, if there is a performance bug with the `sum(y -> ..., ...)` version in the current Julia.

Reply:
Done.